src/main/java/redis/clients/jedis/JedisClusterInfoCache.java

```diff
@@ -49,14 +51,31 @@ public void discoverClusterNodesAndSlots(Jedis jedis) {
       this.nodes.clear();
       this.slots.clear();
 
-      String localNodes = jedis.clusterNodes();
-      for (String nodeInfo : localNodes.split("\n")) {
-        ClusterNodeInformation clusterNodeInfo = nodeInfoParser.parse(nodeInfo, new HostAndPort(
-            jedis.getClient().getHost(), jedis.getClient().getPort()));
+      List<Object> slots = jedis.clusterSlots();
 
-        HostAndPort targetNode = clusterNodeInfo.getNode();
-        setNodeIfNotExist(targetNode);
-        assignSlotsToNode(clusterNodeInfo.getAvailableSlots(), targetNode);
+      for (Object slotInfoObj : slots) {
+        List<Object> slotInfo = (List<Object>) slotInfoObj;
+
+        if (slotInfo.size() <= 2) {
+          continue;
+        }
+
+        List<Integer> slotNums = getAssignedSlotArray(slotInfo);
+
+        // hostInfos
+        int size = slotInfo.size();
+        for (int i = MASTER_NODE_INDEX; i < size; i++) {
+          List<Object> hostInfos = (List<Object>) slotInfo.get(i);
+          if (hostInfos.size() <= 0) {
+            continue;
+          }
+
+          HostAndPort targetNode = generateHostAndPort(hostInfos);
+          setNodeIfNotExist(targetNode);
+          if (i == MASTER_NODE_INDEX) {
```

Comment:
@marcosnils 
For now, assignSlotsToNode() overwrites slot to node information, which means slave information can overwrite master information.
But as I stated we'd be better to store slaves as nodes, so slaves should be added to nodes but without slot assignments.